### PR TITLE
Bluetooth: controller: Fix EVENT_OVERHEAD_START_US jitter 

### PR DESCRIPTION
**Pull request by cvinayak**
As EVENT_OVERHEAD_START_US offset is used in ticks unit in
LLL, ULL scheduling using ticker should also use ticks unit
for EVENT_OVERHEAD_START_US when reducing the first
connection event preparation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c
Comment (andrzej-kaczmarek) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: Why we should subtract TX chain here? `conn_space_us` is calculated in `lll_scan_prepare_connect_req` relative to advertising PDU as "radio RX PDU end time + T_ifs + length of CONNECT_IND". This should be corrected by RX delay of advertising PDU to get actual end of CONNECT_IND over the air.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: The delay to capture timer is 0.6 us (ceiled to 1 us) which is the value in `radio_tx_chain_delay_get(...)`. `radio_rx_chain_delay_get`  is 9.4 us for 1M (ceiled to 10us), which is delay for the Tx bits to be seen on the antenna.

Comment (andrzej-kaczmarek) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: We need to adjust for delay between actual end of PDU over the air and EVENT_END in the hardware which occurs later and captures CC[2] timer - I believe this is delay for RX path we need to adjust here.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: delay in the Rx path is 0.6 us. This is how it has been and passing conformance tests, without reverting most initiator and master tests fail.

Comment (andrzej-kaczmarek) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: > delay in the Rx path is 0.6 us.

No, you can see it here:

image

This was captured on nRF52840 and GPIO is triggered by the same EVENT_END as CC[2] capture.

> This is how it has been and passing conformance tests, without reverting most initiator and master tests fail.

Conformance tests are also failing on 0eb3b3c2c0 which is before extended initiator patches were merged.  For example, here's LL/CON/INI/BV-02-C failing on latest Harmony 2019.65:

```
Test Results:
Validate_TIFS is enabled and 
m_bdAddrs: 0xe2fa3319011d0xc0dec0ffbbaa
Frame #3 : Error - First connection event arrived 5.9 micro-seconds too early
```

image

transmitWindowOffset in CONNECT_IND is 0, so the gap should be at least 1.25ms, so clearly there's incorrect calculation somewhere.

> Since CONNECT_IND is transmitted last, and master transmits first in the connection event, the tx delay towards the antenna are relatively maintain with transmit offset.

Calculation of anchor point is done on RX so it's relative to RXed advertising PDU, not TXed CONNECT_IND PDU.

> The switch implementation ensures that CONNECT_IND is transmitted with the 10 us delay towards the antenna.
> 
> Refer to this line which subtracts the delay towards the antenna to set up tIFS switching:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/41afd838aabd7edd20fdabb4344bea34be11a4f6/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c#L537

10us `rx_chain_delay` here is for RX path (latency between actual RXed PDU end time and EVENT_END), then you have 41us `tx_ready_delay` for TX path which includes radio ramp-up (time from TXEN to actual TX). See line 523 in the same file for TX-TX transition - you have 1us `tx_chain_delay` for TX path of transmitted PDU and then the same 41us `tx_ready_delay` for TX path of subsequent PDU.

Since we are calculating anchor point relative to PDU RX time, we need to use `rx_chain_delay`.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: > 10us rx_chain_delay here is for RX path (latency between actual RXed PDU end time and EVENT_END), then you have 41us tx_ready_delay for TX path which includes radio ramp-up (time from TXEN to actual TX). See line 523 in the same file for TX-TX transition - you have 1us tx_chain_delay for TX path of transmitted PDU and then the same 41us tx_ready_delay for TX path of subsequent PDU.

You are reading the code incorrect, switch to Tx after currently being in Rx where 10 us is used as advanced trigger the PPI for Tx. And in Tx-Tx, there has been a 10 us delay ahead of the first Tx, hence only 1 us is used.

> Test Results:
Validate_TIFS is enabled and 
m_bdAddrs: 0xe2fa3319011d0xc0dec0ffbbaa
Frame #3 : Error - First connection event arrived 5.9 micro-seconds too early

Not sure if +/- 16 us sleep clock accuracy has been considered here. You should see the first transmission between 1234 us and 1266 us.

With the fix in this PR (which puts back the old code), 308 tests on EBQ pass

Comment (andrzej-kaczmarek) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: 16us accuracy makes sense, so this is issue in Harmony.

But that chain delay thing still does not make sense to me. I am either still missing something here or it works by magic.

> You are reading the code incorrect, switch to Tx after currently being in Rx where 10 us is used as advanced trigger the PPI for Tx. And in Tx-Tx, there has been a 10 us delay ahead of the first Tx, hence only 1 us is used. The chain delay function names could be confusing, but that is the names used in the radio timing section of nRF52 product specification, hence had to stick to them.

So if TX-TX already included 10us TX antenna delay on 1st TX, why still add some 1us for 2nd TX instead of just not adding anything?

Also since 1st connection event is calculated relative to RX PDU (it uses time stamp captured on RX), it should be somewhere adjusted for 10us TX antenna delay but I do not really see this anywhere. The only adjustment is for TXEN-to-READY in https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_master.c#L800 but this one actually makes sense.

Then there's also HCTO calculation for slave. It adjusts, I believe, for delay between address in the air and EVENT_ADDRESS here https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_slave.c#L218 and this should be also RX path delay so why it uses `rx_chain_delay` instead of `tx_chain_delay`?

BTW, also found e.g. this comment on DevZone https://devzone.nordicsemi.com/f/nordic-q-a/43979/does-the-nrf52832-has-some-delay-when-the-shortcut-of-end_start-is-enabeled-in-continuous-receiving-what-is-the-meaning-of-rx-chain-delay-does-rx-chain-delay-in-2msps-mode-is-5us-or-9-4us/172721#172721 which states that t_rxchain is time from antenna to digital radio, so basically RX path delay I assume.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: > So if TX-TX already included 10us TX antenna delay on 1st TX, why still add some 1us for 2nd TX instead of just not adding anything?

You are correct, there is no need of using the function for 1us. It is added to compensate for the +/- 1 us jitter by the timer in triggering on PPI. I will do more testing and may send some fine tuning if required.

> Also since 1st connection event is calculated relative to RX PDU (it uses time stamp captured on RX), it should be somewhere adjusted for 10us TX antenna delay but I do not really see this anywhere.

The CONNECT_IND PDU is tx-ed  and the 1st connection event PDU also is a TX for a central, hence, its in theory Tx-Tx, hence both these PDU are relatively 10 us delayed on-air.

Then there's also HCTO calculation for slave. It adjusts, I believe, for delay between address in the air and EVENT_ADDRESS here https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_slave.c#L218 and this should be also RX path delay so why it uses rx_chain_delay instead of tx_chain_delay?

This is probably a bug, `radio_rx_chain_delay_get` is not required, its just listening more which is not necessary. Will fix it, after I do some more testing.

I also, noticed at inside tIFS of 150 us, the `sw_switch` starts Rx early by the amount of `radio_rx_chain_delay_get`, will fix that too.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: Regarding the Tx Delay:
image

A1-A2: is the ADV_IND PDU of 216 us.
B1: is the EVENT_ADDRESS on the receiver side
B2: is the EVENT_END of the received SCAN_REQ on the advertiser.
B1-B2: is the address (rx) to address (rx) between ADV_IND and SCAN_REQ, which is ~366 us  or  (216 + 150), hence when the advertiser has received the end of SCAN_REQ there is no delay.
The delay is in the ADV_IND going out of the advertiser's antenna. Based on PHY used, the bits on-air being 1M (~10us), 2M (~5 us), S2 (25 us) or S8 (30 us) delayed getting out of the antenna of the transmitter.

Comment (andrzej-kaczmarek) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: RX path delay is the same on both devices so spacing between RX events will always be the same as spacing over the air regardless of delay. Here's the same but measured relative to PDU in the air:

image

TX path delay (measured on advertiser) is significantly lower than RX path (measured on scanner). This was captured on nRF52840 with NimBLE, but Zephyr uses the same tasks/events so it should not be much different.

> The CONNECT_IND PDU is tx-ed and the 1st connection event PDU also is a TX for a central, hence, its in theory Tx-Tx, hence both these PDU are relatively 10 us delayed on-air.

CONNECT_IND does not really matter here because time of 1st connection event is calculated relative to received advertising PDU:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c#L148
`radio_tmr_end_get` is a timestamp captured on EVENT_END for advertising PDU. Then expected end of CONNECT_IND is calculated relative to this. To get the in-the-air timestamp it should be adjusted for RX path delay of advertising PDU:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c#L992
And before sending it should be adjusted for TX ramp-up:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_master.c#L800
...and also probably for TX path delay, but perhaps it's already included in the above, not sure about that. Anyway, it's set to 0.6us in Zephyr so negligible. So IMO the current code does proper calculations for radio delays.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: > CONNECT_IND does not really matter here because time of 1st connection event is calculated relative to received advertising PDU:

Yes, the implementation calculates the timestamp from "relative" Rx End, then compensates for transmitting of CONNECT_IND taking into account the ramp up time and the delay on air for the PDU. The first connection event is "relative" to the end of CONNECT_IND when seen on the radio medium by the peripheral.

I will not directly measure the relative toggle measurement in ellisys:

image

See how the address is 8.345 us ahead of the actual address event in the instant timing...
Add your measured 1.615 us, gives the actual delay a peer sees, 9.96 us.

> So IMO the current code does proper calculations for radio delays.

The EBQ conformance tester sees the first packet upto 8-10 us before the 1234 us.

Comment (andrzej-kaczmarek) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: > > CONNECT_IND does not really matter here because time of 1st connection event is calculated relative to received advertising PDU:
> 
> Yes, the implementation calculates the timestamp from "relative" Rx End, then compensates for transmitting of CONNECT_IND taking into account the ramp up time and the delay on air for the PDU. The first connection event is "relative" to the end of CONNECT_IND when seen on the radio medium by the peripheral.

The same relative RX end is used for sw_switch so in both cases we have RX-TX, except delta between packets is different:

- CONNECT_IND starts at "RX end + tifs"
- 1st PDU starts at "RX end + tifs + CONNECT_IND length + offset"

So in both cases we should adjust the same for RX path and then for TX rampup and this is what current code does (i.e. in both cases rx_chain_delay is used).

I do not really see how tx_chain_delay fits here, but since I may be still missing some other timing adjustments that make it works in this scenario I think it would be the best if someone else takes a look on this and approve if ok.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:996]: @andrzej-kaczmarek I have been doing further testing and found a rounding error which has been causing the confusion for me with use of chain delay. I am reverting back to use of rx_chain_delay and will run the conformance tests. Will fix this PR accordingly.

**Review by andrzej-kaczmarek — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**
